rfcs/text/0012_background_sessions.md

```diff
@@ -0,0 +1,291 @@
+additional parameter, the `sessionId`, which will allow search requests to *resume* instead of *restart*. When
+"restoring" a session, it is crucial that search requests are sent with *identical* search request parameters as were
+sent originally. (For example, relative date ranges should be converted to absolute before passing the request to the
+`search` service.)
+
+Applications query data from Elasticsearch on behalf of the end user in a variety of ways:
+
+- Using `SearchSource` (our higher-level abstracted search API, used by Maps, Lens, and most visualizations, exposed by
+the `data` plugin `search` service)
+- Using `search` (our lower-level search API that accepts raw Elasticsearch query DSL, used by Vega and Timelion,
+exposed by the `data` plugin `search` service)
+- Exposing a route that bypasses our `search` services and queries Elasticsearch directly, then manipulates the data
+before returning a response (as in the case of TSVB and most solutions)
+
+The proposed APIs provide the tools necessary to run long search requests, view their status, cancel incomplete search
+requests, and retrieve results, regardless of the mechanism used to query Elasticsearch.
+
+## Background session service
+
+The primary service which will facilitate background sessions is the server-side `session` service.
+
+When a `search` request is handled, and it contains a `sessionId`, the service will query Elasticsearch to see if there
+is already a saved object for that session. If there isn't, then we will forward the search request to Elasticsearch and
+receive an ID specific to that search request. We store that ID in memory, keyed off of a hash of the request parameters
+and the current username.
+
+We will then periodically query Elasticsearch to see if there is a saved object for that session (meaning the user has
+elected to continue the requests in the background). If we find a saved object, then we will update it to contain the
+search ID keyed off of the request hash, and remove it from memory.
+
+Then, in the future when the same `sessionId` is sent with identical request parameters, we will retrieve the saved
+object. Instead of sending all of the request parameters to Elasticsearch (and restarting the search), we will look up
+the search ID, and send that to Elasticsearch (resuming the search). If that specific search request has completed in
+Elasticsearch, the results will be available immediately.
+
+By default, results are stored in Elasticsearch for 5 days. We may consider adding an advanced setting that will allow
+overriding this default.
+
+The background session service will also use the task manager to periodically monitor the status of incomplete
+background sessions. It will query the list of all incomplete sessions, and check the status of each search that is
+executing. If the search requests are all complete, it will update the corresponding saved object to have a `status` of
+`complete`.
```

Comment:
So it won't delete them, just change the status?

Reply:
we need the save object to retrieve search ids later (so we can retrieve stored results) as well as to show the search in the background search manager